Content.Shared/Body/Systems/SharedBodySystem.Parts.cs

```diff
@@ -0,0 +1,359 @@
+    {
+        if (partId == null || !Resolve(partId.Value, ref part, false))
+            yield break;
+
+        foreach (var slot in part.Organs.Values)
+        {
+            if (!TryComp(slot.Child, out OrganComponent? organ))
+                continue;
+
+            yield return (slot.Child.Value, organ);
+        }
+    }
+
+    public IEnumerable<BodyPartSlot> GetPartAllSlots(EntityUid? partId, BodyPartComponent? part = null)
+    {
+        if (partId == null ||
+            !Resolve(partId.Value, ref part, false))
+            yield break;
+
+        foreach (var slot in part.Children.Values)
+        {
+            yield return slot;
+
+            if (!TryComp(slot.Child, out BodyComponent? childPart))
+                continue;
+
+            foreach (var subChild in GetBodyAllSlots(slot.Child, childPart))
+            {
+                yield return subChild;
+            }
+        }
+    }
+
+    public bool CanAttachPart([NotNullWhen(true)] EntityUid? partId, BodyPartSlot slot, BodyPartComponent? part = null)
+    {
+        return partId != null &&
+               slot.Child == null &&
+               Resolve(partId.Value, ref part, false) &&
+               (slot.Type == null || slot.Type == part.PartType) &&
+               _containers.TryGetContainer(slot.Parent, BodyContainerId, out var container) &&
+               container.CanInsert(partId.Value);
+    }
+
+    public virtual bool AttachPart(
+        EntityUid? partId,
+        BodyPartSlot slot,
+        [NotNullWhen(true)] BodyPartComponent? part = null)
+    {
+        if (partId == null ||
+            !Resolve(partId.Value, ref part, false) ||
+            !CanAttachPart(partId, slot, part))
+            return false;
+
+        DropPart(slot.Child);
+        DropPart(partId, part);
+
+        var container = _containers.EnsureContainer<Container>(slot.Parent, BodyContainerId);
+        if (!container.Insert(partId.Value))
+            return false;
+
+        slot.Child = partId;
+        part.ParentSlot = slot;
+
+        if (TryComp(slot.Parent, out BodyPartComponent? parentPart))
+        {
+            part.Body = parentPart.Body;
+        }
+        else if (TryComp(slot.Parent, out BodyComponent? parentBody))
+        {
+            part.Body = parentBody.Owner;
+        }
+        else
+        {
+            part.Body = null;
+        }
+
+        Dirty(slot.Parent);
+        Dirty(partId.Value);
+
+        if (part.Body is { } newBody)
+        {
+            var argsAdded = new BodyPartAddedEventArgs(slot.Id, part);
+
+            // TODO: Body refactor. Somebody is doing it
+            // EntitySystem.Get<SharedHumanoidAppearanceSystem>().BodyPartAdded(Owner, argsAdded);
+            foreach (var component in AllComps<IBodyPartAdded>(newBody).ToArray())
+            {
+                component.BodyPartAdded(argsAdded);
+            }
```

Comment:
can this really not get removed? i thought the only usage was like hands or something

Reply:
Yes, also wanted to save that for a future PR though (by future I mean like half a day after this one is merged)

Reply:
aight

Reply:
Done in https://github.com/space-wizards/space-station-14/pull/12572